rust/crates/sdk/src/stream/monitor_connection.rs

```diff
@@ -0,0 +1,139 @@
+use super::{Stats, StreamError, WebSocketReport};
+
+use crate::{config::Config, stream::establish_connection::try_to_reconnect};
+
+use data_streams_report::feed_id::ID;
+
+use futures::SinkExt;
+use futures_util::StreamExt;
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+use tokio::{
+    net::TcpStream,
+    sync::{broadcast, mpsc, Mutex},
+};
+use tokio_tungstenite::{
+    tungstenite::Message, MaybeTlsStream, WebSocketStream as TungsteniteWebSocketStream,
+};
+use tracing::{error, info, warn};
+
+pub(crate) async fn run_stream(
+    mut stream: TungsteniteWebSocketStream<MaybeTlsStream<TcpStream>>,
+    report_sender: mpsc::Sender<WebSocketReport>,
+    mut shutdown_receiver: broadcast::Receiver<()>,
+    stats: Arc<Stats>,
+    water_mark: Arc<Mutex<HashMap<String, usize>>>,
+    config: Config,
+    feed_ids: Vec<ID>,
+) -> Result<(), StreamError> {
+    let shutdown_flag = Arc::new(AtomicBool::new(false));
+
+    loop {
+        tokio::select! {
+            message = stream.next() => {
+                match message {
+                    Some(Ok(msg)) => {
+                        match msg {
+                            Message::Text(text) => {
+                                info!("Received text message: {}", text);
+                            }
+                            Message::Binary(data) => {
+                                info!("Received new report from Data Streams Endpoint.");
```

Comment:
There's quite a bit of rightward drift on this function, so maybe breaking these big blocks further into other functions would make reading it a bit easier.